File: codes/rust/18.四数之和.rs
/*
 * @lc app=leetcode.cn id=18 lang=rust
 *
 * [18] 四数之和
 */

// @lc code=start
impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut nums = nums;
        nums.sort();
        Self::n_sum(&nums[..], 4, target)
    }

    // 定义一个框架用来应对几数之和的场景
    fn n_sum(nums: &[i32], n: i32, target: i32) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        if nums.is_empty() {
            return res;
        }
        if nums.len() < n as usize {
            return res;
        }
        if n == 2 {
            // 两数之和解法
            let (mut left, mut right) = (0, nums.len() - 1);
            while left < right {
                let sum = nums[left] + nums[right];
                if sum < target {
                    left += 1;
                } else if sum > target {
                    right -= 1;
                } else {
                    let mut tmp = vec![nums[left], nums[right]];
                    res.push(tmp);
                    // 去重
                    while left < right && nums[left] == nums[left + 1] {
                        left += 1;
                    }
                    while left < right && nums[right] == nums[right - 1] {
                        right -= 1;
                    }
                    left += 1;
                    right -= 1;
                }
            }
        } else {
            for i in 0..(nums.len() - n as usize + 1) {
                if i > 0 && nums[i] == nums[i - 1] {
                    continue;
                }
                // 这儿可以进行剪枝，因为nums是提前排序了的，所以:
                // nums[i..i+n]的和一定是最小值，如果target比最小值还小，那么就是没有匹配项，跳过
                // nums[i] + nums[-n + 1..]的和一定是最大值，如果target比这个最大值还大，那么就是没有匹配项，跳过
                let min_sum = Self::sum(&nums[i..i + n as usize]);
                if min_sum > target as i64 {
                    continue;
                }
                let mut last_start = nums.len() - n as usize + 1;
                let max_sum = nums[i] as i64 + Self::sum(&nums[last_start..]);
                if max_sum < target as i64 {
                    continue;
                }
                let sub_res = Self::n_sum(&nums[i + 1..], n - 1, target - nums[i]);
                for mut item in sub_res {
                    item.push(nums[i]);
                    res.push(item);
                }
            }
        }
        res
    }

    fn sum(nums: &[i32]) -> i64 {
        let mut res = 0;
        for num in nums {
            res += num.to_owned() as i64;
        }
        res
    }
}

// @lc code=end

File: codes/golang/18.四数之和.go


func fourSum(nums []int, target int) [][]int {
	sort.Ints(nums)
	// n 为 4，从 nums[0] 开始计算和为 target 的四元组
	return nSumTarget(nums, 4, 0, target)
}

/* 注意：调用这个函数之前一定要先给 nums 排序 */
// n 填写想求的是几数之和，start 从哪个索引开始计算（一般填 0），target 填想凑出的目标和
func nSumTarget(nums []int, n, start, target int) [][]int {
	sz := len(nums)
	res := [][]int{}
	// 至少是 2Sum，且数组大小不应该小于 n
	if n < 2 || sz < n {
		return res
	}
	// 2Sum 是 base case
	if n == 2 {
		// 双指针那一套操作
		lo, hi := start, sz-1
		for lo < hi {
			sum := nums[lo] + nums[hi]
			left, right := nums[lo], nums[hi]
			if sum < target {
				for lo < hi && nums[lo] == left {
					lo++
				}
			} else if sum > target {
				for lo < hi && nums[hi] == right {
					hi--
				}
			} else {
				res = append(res, []int{left, right})
				for lo < hi && nums[lo] == left {
					lo++
				}
				for lo < hi && nums[hi] == right {
					hi--
				}
			}
		}
	} else {
		// n > 2 时，递归计算 (n-1)Sum 的结果
		for i := start; i < sz; i++ {
			sub := nSumTarget(nums, n-1, i+1, target-nums[i])
			for _, arr := range sub {
				// (n-1)Sum 加上 nums[i] 就是 nSum
				arr = append(arr, nums[i])
				res = append(res, arr)
			}
			for i < sz-1 && nums[i] == nums[i+1] {
				i++
			}
		}
	}
	return res
}